### PR TITLE
Check for uninitialized non-nullable field-like events

### LINKED ISSUE #30067: Review UnassignedFieldsWalker follow-up comments
```C#
  // PROTOTYPE(NullableReferenceTypes): Should UnassignedFieldsWalker
    // inherit from DataFlowPassBase<LocalState> directly since it has simpler
    // requirements than DataFlowPass?
    internal sealed class UnassignedFieldsWalker : DataFlowPass
    {
```

```C#

        private void ReportUninitializedNonNullableReferenceTypeFields()
        {
...
            foreach (var member in thisType.GetMembersUnordered())
            {
                if (member.Kind != SymbolKind.Field)
                {
                    // PROTOTYPE(NullableReferenceTypes): Handle events.
                    continue;
                }
                var field = (FieldSymbol)member;
                // PROTOTYPE(NullableReferenceTypes): Can the HasInitializer
                // call be removed, if the body already contains the initializers?
                if (field.IsStatic || HasInitializer(field))
                {
                    continue;
                }
...
```

### PR DESCRIPTION
**Pull request by agocke**
Fixes #30067

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -47447,6 +47468,7 @@ static void M3()
             var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
             // https://github.com/dotnet/roslyn/issues/31018: Report warnings for // 3 and // 4.
             comp.VerifyDiagnostics(
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "E").WithArguments("event", "E").WithLocation(6, 30),
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >Diagnostic [](start = 16, length = 10)

Missing comment.

File: src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs [hunk 2]
```diff
@@ -8,6 +8,46 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics
...
+        E2 += p => {};
+    }
+}";
+            var comp = CreateCompilation(src, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                    // (12,14): warning CS8618: Non-nullable event 'E1' is uninitialized.
+                    //     internal C()
+                    Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "C").WithArguments("event", "E1").WithLocation(12, 14),
+                    // (22,14): warning CS8618: Non-nullable event 'E1' is uninitialized.
+                    //     internal C(object o)
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs]: >internal C(object o) [](start = 27, length = 20)

`E1` is initialized with `+=`.

File: src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs [hunk 3]
```diff
@@ -341,11 +341,13 @@ protected virtual void ReportUnassignedOutParameter(ParameterSymbol parameter, S
...
 
-        protected static bool HasInitializer(FieldSymbol field)
+        protected static bool HasInitializer(Symbol field) => field switch
         {
-            return (field as SourceMemberFieldSymbol)?.HasInitializer == true ||
-                (field as SynthesizedBackingFieldSymbol)?.HasInitializer == true;
-        }
+            SourceMemberFieldSymbol f => f.HasInitializer,
+            SynthesizedBackingFieldSymbol f => f.HasInitializer,
+            SourceFieldLikeEventSymbol e => e.AssociatedEventField?.HasInitializer == true,
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs]: >? [](start = 66, length = 1)

When does a SourceFieldLikeEventSymbol have a null AssociatedEventField?

File: src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs (hunk 2 above)
Comment (agocke) [reply src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs]: Right now we never respect `+=` for accessibility. This PR confirms that behavior.

File: src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs (hunk 3 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs]: `extern`

Comment (agocke) [reply src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs]: I'll add to the test

File: src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs (hunk 2 above)
Comment (agocke) [reply src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs]: Just changed this to implement the += behavior.

**Pull request merged**